web/search.dart

```diff
@@ -0,0 +1,445 @@
+  input.setAttribute('spellcheck', 'false');
+  input.classes.add('tt-input');
+
+  wrapper.append(input);
+
+  var listBox = document.createElement('div');
+  listBox.setAttribute('role', 'listbox');
+  listBox.setAttribute('aria-expanded', 'false');
+  listBox.style.display = 'none';
+  listBox.classes.add('tt-menu');
+
+  var presentation = document.createElement('div');
+  presentation.classes.add('tt-elements');
+
+  listBox.append(presentation);
+
+  wrapper.append(listBox);
+
+  // Set up various search functionality.
+  String highlight(String text, String query) {
+    final sanitizedText = const HtmlEscape().convert(query);
+    return text.replaceAll(
+        query, "<strong class='tt-highlight'>$sanitizedText</strong>");
+  }
+
+  Element createSuggestion(String query, Map<String, dynamic> match) {
+    var suggestion = document.createElement('div');
+    suggestion.setAttribute('data-href', match['href'] as String);
+    suggestion.classes.add('tt-suggestion');
+
+    var suggestionTitle = document.createElement('span');
+    suggestionTitle.classes.add('tt-suggestion-title');
+    suggestionTitle.innerHtml = highlight(
+        '${match['name']} ${(match['type'] as String).toLowerCase()}', query);
+
+    suggestion.append(suggestionTitle);
+
+    if (match['enclosedBy'] != null) {
+      var fromLib = document.createElement('div');
+      fromLib.classes.add('search-from-lib');
+      fromLib.innerHtml =
+          'from ${highlight((match['enclosedBy'] as Map)['name'], query)}';
+
+      suggestion.append(fromLib);
+    }
+
+    suggestion.addEventListener('mousedown', (event) {
+      event.preventDefault();
+    });
+
+    suggestion.addEventListener('click', (event) {
+      if (match['href'] != null) {
+        print('baseHref: $htmlBase');
+        print('match.href: ${match['href']}');
+        print('assign: $htmlBase${match['href']}');
+
+        window.location.assign('$htmlBase${match['href']}');
+        event.preventDefault();
+      }
+    });
+
+    return suggestion;
+  }
+
+  String? storedValue;
+  var actualValue = '';
+  String? hint;
+
+  var suggestionElements = <Element>[];
+  var suggestionsInfo = <Map<String, dynamic>>[];
+  int? selectedElement;
+
+  void setHint(String? value) {
+    hint = value;
+    inputHint.value = value ?? '';
+  }
+
+  void showSuggestions() {
+    if (presentation.hasChildNodes()) {
+      listBox.style.display = 'block';
+      listBox.setAttribute('aria-expanded', 'true');
+    }
+  }
+
+  void hideSuggestions() {
+    listBox.style.display = 'none';
+    listBox.setAttribute('aria-expanded', 'false');
+  }
+
+  void updateSuggestions(String query, List<Map<String, dynamic>> suggestions) {
+    suggestionsInfo = [];
+    suggestionElements = [];
+    presentation.text = '';
+
+    if (suggestions.length < minLength) {
+      setHint(null);
+      hideSuggestions();
+      return;
+    }
+
+    for (var i = 0; i < suggestions.length; i++) {
```

Comment:
Can this be a for-in loop?

Reply:
Happy to, however, I was planning to make a pass through in a follow-on PR to make the code much more Dart like (I avoided doing too much style conversion in this PR in order break the work into smaller chunks).

Reply:
That sounds good.